Feature/multiple ngram bleu

agannon opened the pull request.

stevenbird commented on nltk/translate/bleu_score.py:
```diff
@@ -106,6 +106,16 @@ def sentence_bleu(
+        hypotheses,
+        weights,
+        smoothing_function=None,
+        auto_reweigh=False
```
Is this a typo? Where is this parameter used?

alvations commented on nltk/translate/bleu_score.py:
```diff
@@ -106,16 +106,17 @@ def sentence_bleu(
+def corpus_bleu_multiple_weights(
+        list_of_references,
+        hypotheses,
+        weights_list,
```
There's no need to rename the class here. Also why is the variable `weights_list` more desired than the defaults `weights` that most users of BLEU expect?

alvations replied:
Ah the normal corpus_bleu has been moved down to another function.

alvations commented on nltk/translate/bleu_score.py:
```diff
@@ -106,16 +106,17 @@ def sentence_bleu(
     Calculate a single corpus-level BLEU score (aka. system-level BLEU) for all
-    the hypotheses and their respective references.
+    the hypotheses and their respective references for each weighting tuple in
+    weights_list
```
This is rather non-normal use of BLEU for >4-grams BLEU, users can already add to the weights list in the original implementation and change the n order.

alvations replied:
Ah the normal corpus_bleu has been moved down to another function.

alvations commented:
Thanks @agannon for the PR. I think this PR is meant to resolve the issue where you want the output to return the fine-grained scores for BLEU. Is that right? 

~The PR's approach to nesting `bleu -> corpus_bleu -> corpus_bleu_multiple_weights` is rather nasty and spaghetti-like. Maybe there's a better approach to design the API for BLEU.~

But I can't think of a better API off my head now.

agannon commented:
> Thanks @agannon for the PR. I think this PR is meant to resolve the issue where you want the output to return the fine-grained scores for BLEU. Is that right?
> But I can't think of a better API off my head now.

Yes, based on this issue: https://github.com/nltk/nltk/issues/2320.

I wanted to avoid changing the type signatures to not disrupt how users already call `corpus_bleu`, but I can see how having 2 functions would be confusing down the line. In terms of a different API do you think it worthwhile to have one function `corpus_bleu` with its type signature as:
```
def corpus_bleu(list_of_references: List[List[str]],
                           hypotheses: List[str],
                           weights: Union[Tuple[float], List[Tuple[float]]],
                           smoothing_function: SmoothingFunction = None,
                           auto_reweigh: bool = False) ->
                           Union[float, List[float]]
```
where weights can either be the original tuple of values OR a list of tuples to consider, and the return is either a single float or a list of corresponding floats.
That way users have the flexibility of using either or.

agannon committed:
Adding myself (agannon) to contributors

agannon commented:
Hi @alvations, I refactored my original changes to incorporate them into one single function. Please let me know if you think this is a better API than before

Edit: I am now seeing that it failed on py3.5 so I will look into that.

agannon committed:
Implements multiple ngram BLEU score function

https://github.com/nltk/nltk/issues/2320

corpus_bleu function runs inefficiently when being used with
different weightings by recalculating the underlying values
each time the function is called instead of reusing them.

* Creates a more generalized function that can calculate BLEU scores
  for multiple pre-specified weightings
* Adjusts other functions/tests impacted accordingly

agannon commented:
Hi @alvations , could you let me know if this API is more to your liking than the nested corpus_bleu functions from earlier? I want to make sure I'm on the right track before I rebase to add the most recent nltk/develop commits to this PR

alvations commented:
@agannon sorry for the delay, let me look at this in the next couple of days and get back to you.

stevenbird committed:
Merge branch 'develop' into feature/multiple-ngram-bleu
